contracts/contracts/Umbra.sol

```diff
@@ -198,11 +209,82 @@ contract Umbra is Ownable {
     bytes32 _r,
     bytes32 _s
   ) external {
+    _validateWithdrawSignature(_stealthAddr, _acceptor, _tokenAddr, _sponsor, _sponsorFee, UmbraHookable(0), "", _v, _r, _s);
+    _withdrawTokenInternal(_stealthAddr, _acceptor, _tokenAddr, _sponsor, _sponsorFee, UmbraHookable(0), "");
+  }
+
+  /**
+   * @notice Withdraw an ERC20 token payment on behalf of a stealth address via signed authorization
+   * @param _stealthAddr The stealth address whose token balance will be withdrawn
+   * @param _acceptor Address where withdrawn funds should be sent
+   * @param _tokenAddr Address of the ERC20 token being withdrawn
+   * @param _sponsor Address which is compensated for submitting the withdrawal tx
+   * @param _sponsorFee Amount of the token to pay to the sponsor
+   * @param _hook Contract that will be called after the token withdrawal has completed
+   * @param _data Arbitrary data that will be passed to the post-withdraw hook contract
+   * @param _v ECDSA signature component: Parity of the `y` coordinate of point `R`
+   * @param _r ECDSA signature component: x-coordinate of `R`
+   * @param _s ECDSA signature component: `s` value of the signature
+   */
+  function withdrawTokenAndCallOnBehalf(
+    address _stealthAddr,
+    address _acceptor,
+    address _tokenAddr,
+    address _sponsor,
+    uint256 _sponsorFee,
+    UmbraHookable _hook,
+    bytes memory _data,
+    uint8 _v,
+    bytes32 _r,
+    bytes32 _s
+  ) external {
+    _validateWithdrawSignature(_stealthAddr, _acceptor, _tokenAddr, _sponsor, _sponsorFee, _hook, _data, _v, _r, _s);
+    _withdrawTokenInternal(_stealthAddr, _acceptor, _tokenAddr, _sponsor, _sponsorFee, _hook, _data);
+  }
+
+  /// @dev low level withdrawal function that should only be called after safety checks
+  function _withdrawTokenInternal(
+    address _stealthAddr,
+    address _acceptor,
+    address _tokenAddr,
+    address _sponsor,
+    uint256 _sponsorFee,
+    UmbraHookable _hook,
+    bytes memory _data
+  ) internal {
     uint256 _amount = tokenPayments[_stealthAddr][_tokenAddr];
 
-    // also protects from underflow
+    // also protects from underflow and re-entrance
     require(_amount > _sponsorFee, "Umbra: No balance to withdraw or fee exceeds balance");
 
+    uint256 _withdrawalAmount = _amount - _sponsorFee;
+    delete tokenPayments[_stealthAddr][_tokenAddr];
+    emit TokenWithdrawal(_stealthAddr, _acceptor, _withdrawalAmount, _tokenAddr);
+
+    SafeERC20.safeTransfer(IERC20(_tokenAddr), _acceptor, _withdrawalAmount);
+
+    if (_sponsorFee > 0) {
+      SafeERC20.safeTransfer(IERC20(_tokenAddr), _sponsor, _sponsorFee);
+    }
+
+    if (address(_hook) != address(0)) {
+      _hook.callHook(_withdrawalAmount, _stealthAddr, _acceptor, _tokenAddr, _sponsor, _sponsorFee, _data);
```

Comment:
I considered this. I left it out because of gas considerations, and because the hook itself could always emit an event if it needed something recorded. Do you think it's worth gas expenses?

Reply:
> the hook itself could always emit an event if it needed something recorded

True, though that doesn't help much with general chain analytics to see, e.g. what % of withdraws call a hook

Maybe we add this to the discussion as part of #119?

Reply:
Yeah, I agree it would be useful to some extent, but it's the same as indexing the params: is the use worth the gas costs. Cool, added to that PR for further investigation.